src/sequence_batch_scheduler.cc

```diff
@@ -110,12 +108,143 @@ SequenceBatchScheduler::Create(
   // runner. This is at least 1 even if the model doesn't support
   // batching.
   const size_t model_batch_size = std::max(1, config.max_batch_size());
-  size_t seq_slot_cnt = model_batch_size;
+  sched->seq_slot_cnt_ = model_batch_size;
   if (config.sequence_batching().has_oldest()) {
-    seq_slot_cnt =
+    sched->seq_slot_cnt_ =
         config.sequence_batching().oldest().max_candidate_sequences();
   }
 
+  // Create batchers.
+  RETURN_IF_ERROR(sched->CreateBatchers(model->BackgroundInstances()));
+
+  // Create a reaper thread that watches for idle sequences. Run the
+  // reaper a lower priority.
+  SequenceBatchScheduler* raw = sched.release();
+
+  raw->reaper_thread_exit_ = false;
+  raw->reaper_thread_.reset(
+      new std::thread([raw]() { raw->ReaperThread(10 /* nice */); }));
+
+  scheduler->reset(raw);
+
+  return Status::Success;
+}
+
+Status
+SequenceBatchScheduler::Update()
+{
+  // Find added and removed instances for this update.
+  std::vector<std::shared_ptr<TritonModelInstance>> added_instances;
+  std::unordered_set<TritonModelInstance*> removed_instances;
+  InstancesDiff(&added_instances, &removed_instances);
+
+  // Batchers should be destroyed outside the lock.
+  std::vector<std::unique_ptr<SequenceBatch>> removed_batchers;
+
+  {
+    std::unique_lock<std::mutex> lk(mu_);
+
+    // Add the new batchers.
+    RETURN_IF_ERROR(CreateBatchers(added_instances));
+
+    // All sequence slots of the removed instances are pending to be removed.
+    pending_removal_seq_slots_.clear();
+    for (auto& instance : removed_instances) {
+      pending_removal_seq_slots_.emplace(instance, seq_slot_cnt_);
+    }
+
+    // Erase ready slots which its instance is pending removal.
+    std::priority_queue<
+        BatcherSequenceSlot, std::vector<BatcherSequenceSlot>,
+        BatcherSequenceSlotCompare>
+        new_ready_batcher_seq_slots;
+    while (!ready_batcher_seq_slots_.empty()) {
+      auto& ready_batcher_seq_slot = ready_batcher_seq_slots_.top();
+      if (removed_instances.find(ready_batcher_seq_slot.model_instance_) ==
+          removed_instances.end()) {
+        // The ready slot is not being removed.
+        new_ready_batcher_seq_slots.push(ready_batcher_seq_slot);
+      } else {
+        // The ready slot is removed, so update pending removal instances.
+        ErasePendingRemovalSequenceSlot(ready_batcher_seq_slot);
+      }
+      ready_batcher_seq_slots_.pop();
+    }
+    ready_batcher_seq_slots_.swap(new_ready_batcher_seq_slots);
+
+    // Wait until all removed batchers have completed its assigned sequence.
+    pending_removal_seq_slots_cv_.wait(
+        lk, [this] { return pending_removal_seq_slots_.empty(); });
```

Comment:
So we lose a model loading thread here waiting for sequences to be finished?

Reply:
Yes. The model loading thread will be paused, waiting for sequences belonging to the instances that will be removed to finish.

Reply:
So if there are 4 sequence model to be updated and each has 1 active sequences there is a client keep sending requests to keep all the sequences alive. We can't perform any model update as we only have 4 model loading threads by default?

If so, can this be something passive that the scheduler itself will remove the instance when it is done with ongoing sequences? `Update()` here will just mark the instances to be removed if it can't be done immediately?

Reply:
> if there are 4 sequence model to be updated and each has 1 active sequences there is a client keep sending requests to keep all the sequences alive. We can't perform any model update as we only have 4 model loading threads by default?

Yes, but I think the user may specify more model loading threads in such case.

> can this be something passive that the scheduler itself will remove the instance when it is done with ongoing sequences? Update() here will just mark the instances to be removed if it can't be done immediately?

I think it is doable, because the shared ptr of the removed instances can be held by the update function and passed to a background thread.

A point worth noting is test cases involving any sequence instance removal will need to poll for the final removal count, as instances can be removing after the update api returns.

Reply:
The implementation is added, see [Remove batcher pending sequence completion after update return](https://github.com/triton-inference-server/core/pull/207/commits/93f46109d1732e354b568780de0eb475a51d37ab)
, [[WIP] Async destruction of batchers](https://github.com/triton-inference-server/core/pull/207/commits/5524945f4e48b01e52384c0a7371acbc65ef91b6)
 and [Async destruction of resources](https://github.com/triton-inference-server/core/pull/207/commits/00d0ef13e8a10dbb599e4c8f573d2422a3e6c96b)